### PR TITLE
feat(schematics): list schematics from extended collections

### PR DESCRIPTION
**Pull request by xdissent**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

The CLI lists available schematics from custom collections by parsing the `collection.json` file directly. This list does not include schematics that may be available via an `extends` key in the `collection.json`. It also looks for `collection.json` in a fixed location, rather than supporting the `schematics` key of `package.json`.

Issue Number: N/A

## What is the new behavior?

The CLI lists *all* available schematics (with `extends` support) and locates the `collection.json` using the `schematics` key of `package.json`. Additionally, schematic names and aliases are deduped and merged to better reflect the actual configuration that will be used by `@angular-devkit/schematics`: Collections using `extends` can overwrite extended collection schematics and re-map aliases. An alias may also override a schematic name, causing it to be effectively excluded from the collection.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

Has allowing multiple schematic aliases in nest-cli been considered? I think it would make it more clear what is going on with how `@angular-devkit/schematics` merges schematic names and aliases. Right now there are hidden aliases defined by the schematics that can be used but not listed by the CLI.

### CODE REVIEW

File: lib/schematics/custom.collection.ts [hunk 1]
```diff
@@ -11,18 +10,26 @@ export interface CollectionSchematic {
...
+    ];
+    const usedNames = new Set<string>();
+    const schematics: Schematic[] = [];
+    for (const collectionDesc of collectionDescs) {
+      const schematicsDescs = Object.entries(collectionDesc.schematics);
+      for (const [name, { description, aliases = [] }] of schematicsDescs) {
+        if (usedNames.has(name)) continue;
+        usedNames.add(name);
+        const alias = aliases.find((a) => !usedNames.has(a)) ?? name;
+        for (const alias of aliases) usedNames.add(alias);
```
Comment (kamilmysliwiec) [inline lib/schematics/custom.collection.ts:27]:
```suggestion
        for (const alias of aliases) {
           usedNames.add(alias);
        }
```

File: lib/schematics/custom.collection.ts [hunk 2]
```diff
@@ -11,18 +10,26 @@ export interface CollectionSchematic {
...
+    const collectionDescs = [
+      collection.description,
+      ...(collection.baseDescriptions ?? []),
+    ];
+    const usedNames = new Set<string>();
+    const schematics: Schematic[] = [];
+    for (const collectionDesc of collectionDescs) {
+      const schematicsDescs = Object.entries(collectionDesc.schematics);
+      for (const [name, { description, aliases = [] }] of schematicsDescs) {
+        if (usedNames.has(name)) continue;
```
Comment (kamilmysliwiec) [inline lib/schematics/custom.collection.ts:24]:
```suggestion
        if (usedNames.has(name)) {
            continue;
        }
```

**Comment by kamilmysliwiec** [comment]
Tests seem to fail for this PR for some reason

**Comment by micalevisk** [comment]
@xdissent did you got the unit tests passing in your machine? `npm run test`

**Comment by xdissent** [comment]
Sorry, I was on vacation. Yep, they passed locally. I’ll rebase and try again soon.

**Comment by johnny9144** [comment]
hey, guys, thanks for working on this, this pull request not only allowed us to list properly available schematics but fixed the bug that caused the `nest` command failure when using a relative path of `collection.json` as well as a package that doesn't put `collection.json` in the project root.

I took some time to dig into why the tests failed, I don't think it was caused by the changes, it's because of the unstable test cases in `test/lib/compiler/hooks/tsconfig-paths.hook.spec.ts`. @kamilmysliwiec has made it skip in this [commit](https://github.com/nestjs/nest-cli/commit/f6264abedb6f59bb3229dcfdc583a606cc7408f3). 

so, I think we can fix the test cases simply by rebasing the master. @xdissent

**Comment by xdissent** [comment]
Sorry that took so long, I started a new full-time gig right after I got back from vacation 😅

**Review by rubiin — APPROVED** [review]

File: lib/schematics/custom.collection.ts [hunk 3]
```diff
@@ -11,18 +10,30 @@ export interface CollectionSchematic {
...
-    const schematics = Object.entries(collection.schematics).map(
-      ([name, value]) => {
-        const schematic = value as CollectionSchematic;
-        const description = schematic.description;
-        const alias = schematic?.aliases?.length ? schematic.aliases[0] : '';
-        return { name, description, alias };
-      },
+    const workflow = new NodeWorkflow(process.cwd(), {});
+    const collection = workflow.engine.createCollection(this.collection);
+    const collectionDescs = [
```
Comment (kamilmysliwiec) [inline lib/schematics/custom.collection.ts:15]:
```suggestion
    const collectionDescriptions = [
```

Comment (kamilmysliwiec) [reply lib/schematics/custom.collection.ts:15]: nit

**Pull request merged**